components/net/fetch/methods.rs

```diff
@@ -90,14 +91,179 @@ pub fn fetch(request: Rc<Request>, cors_flag: bool) -> Response {
     // TODO: Figure out what a Priority object is
     // Step 3
     // Step 4
-    main_fetch(request, cors_flag)
+    main_fetch(request, cors_flag, false)
 }
 
 /// [Main fetch](https://fetch.spec.whatwg.org/#concept-main-fetch)
-fn main_fetch(request: Rc<Request>, _cors_flag: bool) -> Response {
+fn main_fetch(request: Rc<Request>, cors_flag: bool, recursive_flag: bool) -> Response {
     // TODO: Implement main fetch spec
-    let response = basic_fetch(request);
-    response
+
+    // Step 1
+    let mut response = None;
+
+    // Step 2
+    if request.local_urls_only {
+        match &*request.current_url().scheme {
+            "about" | "blob" | "data" | "filesystem" => response = Some(Response::network_error()),
+            _ => { }
+        };
+    }
+
+    // Step 3
+    // TODO be able to execute report CSP
+
+    // Step 4
+    // TODO this step, based off of http_loader.rs
+
+    // Step 5
+    // TODO this step
+
+    // Step 6
+    if request.referer == Referer::NoReferer {
+        // TODO be able to invoke "determine request's referer"
+    }
+
+    // Step 7
+    // TODO this step
+
+    // Step 8
+    if request.synchronous && !recursive_flag {
+        // TODO run the remaining steps in parallel
+    }
+
+    // Step 9
+    let mut response = if response.is_none() {
+
+        let current_url = request.current_url();
+        let origin_match = match request.origin {
+            Some(ref origin) => origin.origin() == current_url.origin(),
+            _ => false
+        };
+
+        if (!cors_flag && origin_match) || (current_url.scheme == "data" && request.same_origin_data.get()) ||
+            current_url.scheme == "about" || request.mode == RequestMode::Navigate {
+            basic_fetch(request.clone())
+
+        } else if request.mode == RequestMode::SameOrigin {
+            Response::network_error()
+
+        } else if request.mode == RequestMode::NoCORS {
+            request.response_tainting.set(ResponseTainting::Opaque);
+            basic_fetch(request.clone())
+
+        } else if current_url.scheme == "http" || current_url.scheme == "https" {
+            Response::network_error()
+
+        } else if request.use_cors_preflight || request.unsafe_request &&
+            (!is_simple_method(&request.method.borrow()) ||
+            request.headers.borrow().iter().any(|h| !is_simple_header(&h))) {
+
+            request.response_tainting.set(ResponseTainting::CORSTainting);
+            request.redirect_mode.set(RedirectMode::Error);
+            let response = http_fetch(request.clone(), BasicCORSCache::new(), cors_flag, true, false);
+            if Response::is_network_error(&response) {
+                // TODO clear cache entries using request
+            }
+            response
+
+        } else {
```

Comment:
I have a strong feeling that we can do some magic with a match-expression with a tuple.